### PR TITLE
Add basic metadata importing to init command

### LINKED ISSUE #191: Import requirements.txt / setup.py Projects
It's quite hard to get started with rye today if one does not have a pyproject.toml project. A possible way would be to have a `rye init --import` command that takes an existing project and makes it work with rye.

* #103

### PR DESCRIPTION
**Pull request by cnpryer**
## Summary of changes

- Add [`configparser`](https://crates.io/crates/configparser) for INI file parsing
- Add  [`monotrail-utils`](https://github.com/konstin/poc-monotrail/) for `RequirementsTxt` requirements file parsing.
- `rye init` by default will attempt an import from setup.cfg and setup.py files (setup.py is prioritized)
- Add `--no-import` argument
- Add `-r`/`--requirements` and `--dev-requirements` arguments for requirements files

## How it works

If `--no-import` is not used:
 1. Attempt to read metadata from setup.py
 1. Attempt to read from setup.cfg
 1. Attempt to read requirements from `-r`/`--requirements`/`--dev-requirements`

- [x] `RequirementsTxt` is used from [`monotrail-utils`](https://github.com/konstin/poc-monotrail/) for requirements file parsing.
  - https://github.com/konstin/poc-monotrail/pull/53
    - See [unsupported features](https://github.com/konstin/poc-monotrail/blob/e6526ad3dd9700a05b5a6b3708ee12a8c54fe017/monotrail-utils/src/requirements_txt.rs#L12-L16).
    - No license.
- [x] If setup.py/setup.cfg is found or requirements are provided `rye` attempts an import.
    - Can skip with `--no-import`

Tag #191

----
TODO:
- [x] import setup.cfg first-pass
- [x] import setup.py first-pass
- [x] import requirements first-pass (see [unsupported features](https://github.com/konstin/poc-monotrail/blob/e6526ad3dd9700a05b5a6b3708ee12a8c54fe017/monotrail-utils/src/requirements_txt.rs#L12-L16))

### CODE REVIEW

**Comment by cnpryer** [comment]
> One thought I have is that if `--import` is not passed, but there is something importable it should probably not do something stupid but let the user know that they can import instead.

Yea that'd be nice. We could: 

a. prompt the user and wait for a response (I'm not in love with this approach)
b. emit the notice and allow the user to re-do it (Not in love with this either)
c. assume an import if either file exists (added in e18f12d3ca796824d14cf35cdf2ae1e680142589)

There's probably a bunch more QoL stuff we can bake into project initialization like handling existing partial pyproject.toml files. Feel free to throw more ideas at me if you think of any others.

Not sure if I'll add everything here, but I can throw them in new PRs if needed.

**Comment by cnpryer** [comment]
> c. assume an import if either setup.py or setup.cfg file exists (added in https://github.com/mitsuhiko/rye/commit/e18f12d3ca796824d14cf35cdf2ae1e680142589)

It might be nice to just assume an import silently without any `--import` flag regardless. Requirements flags can still be used to add the requirements file imports on top of it. But I'll leave the flag in for now to let use chew on it more.

**Comment by cnpryer** [comment]
> One thought I have is that if `--import` is not passed, but there is something importable it should probably not do something stupid but let the user know that they can import instead.

e18f12d3ca796824d14cf35cdf2ae1e680142589 adds an assumed import. If you'd prefer waiting for user input or something similar/more explicit I'll modify it. 

I'm kind of torn on this because I know that a better experience is probably fully supporting both setup.py and setup.cfg files or committing to pyproject.toml-managed projects with `--import` as a way to migrate over.

File: rye/src/cli/init.rs [hunk 1]
```diff
@@ -282,3 +353,205 @@ pub fn execute(cmd: Args) -> Result<(), Error> {
...
+    requires_python: Option<String>,
+    license: Option<String>,
+    dependencies: Option<Vec<String>>,
+    dev_dependencies: Option<Vec<String>>,
+}
+
+/// Check if setup.py or setup.cfg exist.
+fn import_available<T: AsRef<Path>>(dir: T) -> bool {
+    let dir = dir.as_ref();
+    dir.join("setup.py").is_file() | dir.join("setup.cfg").is_file()
```
Comment (mitsuhiko) [inline rye/src/cli/init.rs:372]: I'm not a fan of `|` instead of `||`. Same for the other one.

Comment (cnpryer) [reply rye/src/cli/init.rs:372]: Oh cool `||` is a lazy operator. TIL

Comment (cnpryer) [reply rye/src/cli/init.rs]: 083de3442807e26ccb1176f0b2501e0419b4eb59

File: rye/src/cli/init.rs [hunk 2]
```diff
@@ -75,7 +81,15 @@ authors = [
...
-dependencies = {{ dependencies }}
+{%- if dependencies %}
+dependencies = [
+{%- for dependency in dependencies %}
+    {{ dependency }},
+{%- endfor %}
+]
+{%- else %}
+dependencies = []
+{%- endif %}
```
Comment (cnpryer) [inline rye/src/cli/init.rs]: Also added formatting for dependencies

File: rye/src/cli/init.rs [hunk 3]
```diff
@@ -101,8 +115,14 @@ build-backend = "pdm.backend"
...
-{%- if dev_dependencies is not none %}
-dev-dependencies = {{ dev_dependencies }}
+{%- if dev_dependencies %}
+dev_dependencies = [
+{%- for dependency in dev_dependencies %}
+    {{ dependency }},
+{%- endfor %}
+]
+{%- else %}
+dev_dependencies = []
```
Comment (cnpryer) [inline rye/src/cli/init.rs:125]: Added formatting here too

**Comment by cnpryer** [comment]
I'm changing this PR to only tag #191 instead of closing. I think we'd want to close it when/if requirements parsing is fully supported, when we're able to import from poetry.coml (or other managed projects), and after we'd decided if we see a future where Rye fully supports setup file managed projects.

**Comment by cnpryer** [comment]
Marking this as ready for a review, but I'd like to do more testing later. Here's where it's at so far:

See PR description for callouts.
1. We use the `monotrail-utils` crate for requirements parsing.
2. We perform the import when `rye init` is called and a setup.py or setup.cfg is found
3. You cant import all supported requirements formats (see [this link](https://github.com/konstin/poc-monotrail/blob/4bf56cb89a481fe4dcea4308028cb3a4047c1c81/src/requirements_txt.rs#L12)).

File: rye/src/cli/init.rs [hunk 4]
```diff
@@ -291,3 +393,244 @@ pub fn execute(cmd: Args) -> Result<(), Error> {
...
+    }
+    if let Some(section) = config.get("options") {
+        if metadata.requires_python.is_none() {
+            if let Some(Some(python_requires)) = section.get("python_requires") {
+                metadata.requires_python = Some(python_requires.to_string());
+            }
+        }
+        if let Some(Some(reqs)) = section.get("install_requires") {
+            reqs.lines()
+                .filter_map(|x| Requirement::from_str(x).ok())
```
Comment (cnpryer) [inline rye/src/cli/init.rs:582]: This will silently fail

File: rye/src/cli/init.rs [hunk 5]
```diff
@@ -291,3 +393,244 @@ pub fn execute(cmd: Args) -> Result<(), Error> {
...
+        .output()?;
+    if cmd.status.success() {
+        Ok(serde_json::from_slice(&cmd.stderr)?)
+    } else {
+        let log = String::from_utf8_lossy(&cmd.stderr);
+        bail!("failed to proxy setup.py\n{}", log);
+    }
+}
+
+/// See https://github.com/konstin/poc-monotrail/blob/7487250e5ace3447f25a5573b7a9953ccd9537e/src/requirements_txt.rs#L12-L16
```
Comment (ischaojie) [inline rye/src/cli/init.rs:620]: It looks like the link here has a 404 status.

Comment (cnpryer) [reply rye/src/cli/init.rs:620]: Nice catch

File: rye/src/cli/init.rs [hunk 6]
```diff
@@ -291,3 +393,252 @@ pub fn execute(cmd: Args) -> Result<(), Error> {
...
+
+fn import_setup_py<T: AsRef<Path>>(
+    metadata: &mut Metadata,
+    requirements: &mut BTreeMap<String, String>,
+    path: T,
+    python: T,
+) -> Result<(), Error> {
+    let json = get_setup_py_json(path, python)?;
+    if let Some(Value::String(name)) = json.get("name") {
+        if metadata.name.is_none() {
```
Comment (cnpryer) [inline rye/src/cli/init.rs:488]:
```suggestion
    if metadata.name.is_none() {
        if let Some(Value::String(name)) = json.get("name") {
```
Might flip these

**Comment by cnpryer** [comment]
Have you thought about how we want to do testing in general? I'm happy to introduce some here for this change, but I want to align with broader testing of rye.

**Comment by ischaojie** [comment]
I've been concerned about the lack of extensive testing for rye, and if someone can add some tests as a reference (I lack experience in this area of rust), I'd be happy to contribute.

**Comment by mitsuhiko** [comment]
Rye definitely has reached the point where it needs tests. I have started something a while back based on [insta-cmd](https://docs.rs/insta-cmd/latest/insta_cmd/) but it was all quite dependent on external sources and felt super fragile.

**Comment by mitsuhiko** [comment]
I will merge this now. I did not make progress on tests yet, but that should not hold this up.

**Pull request merged**